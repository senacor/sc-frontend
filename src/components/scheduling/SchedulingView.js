import React from 'react';
import Typography from '@material-ui/core/Typography';
import Grid from '@material-ui/core/Grid';
import * as actions from '../../actions';
import { connect } from 'react-redux';
import moment from 'moment';
import {
  getAppointments,
  getPrDetail,
  getSelectedDate,
  getMeeting
} from '../../reducers/selector';
import PersonToggle from './PersonToggle';
import TimeTable from './AppointmentTable/TimeTable';
import Attendee from './AppointmentTable/Attendee';
import { extractAppointments } from './AppointmentTable/AppointmentUtilities';
import ObjectGet from 'object-get';
import MeetingDetailsView from './MeetingDetailsView';
import MeetingCreatorForm from './MeetingCreatorForm';

export class SchedulingView extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      reviewer: ''
    };

    this.state = {
      employee: {
        id: props.prDetail.employee.login,
        name:
          ObjectGet(props, 'prDetail.employee.firstName') +
          ' ' +
          ObjectGet(props, 'prDetail.employee.lastName'),
        role: 'Ich',
        show: false
      }
    };
    if (
      props.prDetail.reviewer !== undefined &&
      props.prDetail.reviewer.id !== ''
    ) {
      this.state.reviewer = {
        id: props.prDetail.reviewer.login,
        name:
          ObjectGet(props, 'prDetail.reviewer.firstName') +
          ' ' +
          ObjectGet(props, 'prDetail.reviewer.lastName'),
        role: 'Beurteiler',
        show: false
      };
    }
    if (
      props.prDetail.supervisor !== undefined &&
      props.prDetail.supervisor.id !== ''
    ) {
      this.state.supervisor = {
        id: props.prDetail.supervisor.login,
        name:
          ObjectGet(props, 'prDetail.supervisor.firstName') +
          ' ' +
          ObjectGet(props, 'prDetail.supervisor.lastName'),
        role: 'Vorgesetzter',
        show: false
      };
    }
  }

  render() {
    const allAppointments = this.props.appointmentsSearchResults;
    const { meeting } = this.props;
    return (
      <div id={'outer'}>
        {meeting == null ? (
          <React.Fragment>
            <Typography gutterBottom variant="display1">
              Terminfindung
            </Typography>
            <Grid
              id={'tableRolePick'}
              container
              spacing={24}
              direction="column"
            >
              <Grid item>
                <MeetingCreatorForm
                  prById={this.props.prDetail}
                  fetchAppointments={this.fetchAppointments}
                />
              </Grid>
              <Grid item>
                <Grid container direction="column">
                  {Object.getOwnPropertyNames(this.state).map(attendee => {
                    return (
                      <Grid item key={attendee}>
                        <PersonToggle
                          displayName={`${this.state[attendee].id}`}
                          displayRole={`${this.state[attendee].role}`}
                          onChange={this.onVisibilityChange(attendee)}
                          showAttendee={this.state[attendee].show}
                        />
                      </Grid>
                    );
                  })}
                </Grid>
              </Grid>
              <Grid item>
                <TimeTable>
                  {Object.getOwnPropertyNames(this.state)
                    .filter(
                      key =>
                        this.state[key] &&
                        this.state[key].id &&
                        allAppointments[this.state[key].id]
                    )
                    .map((attendee, index, keyArray) => (
                      <Attendee
                        key={attendee}
                        show={this.state[attendee].show}
                        appointments={extractAppointments(
                          allAppointments[this.state[attendee].id].appointments
                        )}
                        selectedDate={this.props.selectedDate}
                        distanceFromLeft={(100 * index) / keyArray.length}
                      />
                    ))}
                </TimeTable>
              </Grid>
            </Grid>
          </React.Fragment>
        ) : (
          <MeetingDetailsView />
        )}
      </div>
    );
  }

  componentDidMount() {
    this.props.fetchMeeting(this.props.prDetail);
    this.fetchAppointments(
      moment()
        .local()
        .format('YYYY-MM-DD')
    );
  }

  fetchAppointments = date => {
    this.props.appointmentsSearch(
      Object.getOwnPropertyNames(this.state)
        .map(role => this.state[role].id)
        .join(),
      date
    );
  };

  onVisibilityChange = attendee => () => {
    this.setState({
      [attendee]: Object.assign({}, this.state[attendee], {
        show: !this.state[attendee].show
      })
    });
  };
}

export default connect(
  state => ({
    appointmentsSearchResults: getAppointments(state),
    prDetail: getPrDetail()(state),
    selectedDate: getSelectedDate(state),
    meeting: getMeeting(state)
  }),
  {
    appointmentsSearch: actions.appointmentsSearch,
    fetchMeeting: actions.fetchMeeting
  }
)(SchedulingView);
