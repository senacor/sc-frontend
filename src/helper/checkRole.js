import ROLES from './roles';

export const hasRole = (role, userroles) => {
  return userroles.includes(role);
};

export const isEmployee = (userroles = []) => {
  return userroles.includes('PR_Mitarbeiter');
};

export const isSupervisor = userroles => {
  return userroles.includes(ROLES.PR_CST_LEITER);
};
