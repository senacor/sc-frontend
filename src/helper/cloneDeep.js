const cloneDeep = require('lodash/cloneDeep');

export default cloneDeep;
