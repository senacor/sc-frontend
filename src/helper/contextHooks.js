import { useContext } from 'react';
import { ErrorContext, InfoContext, UserinfoContext } from '../components/App';
import ROLES from './roles';

export const useInfoContext = () => {
  const infoContext = useContext(InfoContext.context);
  return {
    context: infoContext,
    msg: msg => {
      infoContext.setValue({
        hasInfos: true,
        messageId: msg
      });
    },
    hide: () => {
      infoContext.setValue({
        hasInfos: false,
        messageId: ''
      });
    }
  };
};

export const useUserinfoContext = () => {
  const userInfoContext = useContext(UserinfoContext.context);
  const { userroles, userinfo } = userInfoContext.value;
  return {
    context: userInfoContext,
    isSupervisorInPr: pr => {
      return userinfo.userId === pr.supervisor.id;
    },
    isReviewerInPr: pr => {
      return userinfo.userId === pr.reviewer.id;
    },
    isOwnerInPr: pr => {
      return pr.employee.id === userinfo.userId;
    },
    hasRoleAdmin: () => {
      return userroles[0] === 'ADMIN';
    },
    hasRoleHr: () => {
      return userroles.includes(ROLES.PERSONAL_DEV);
    },
    hasRoleEmployee: () => {
      return userroles.includes(ROLES.DEVELOPER);
    },
    hasRoleSupervisor: () => {
      return userroles.includes(ROLES.SUPERVISOR);
    }
  };
};

export const useErrorContext = () => {
  const errorContext = useContext(ErrorContext.context);
  return {
    context: errorContext,
    showGeneral: () => {
      errorContext.setValue({
        hasErrors: true,
        messageId: 'message.error'
      });
    },
    hide: () => {
      errorContext.setValue({ hasErrors: false, messageId: '' });
    },
    hasError: error => {
      return errorContext.value.errors && errorContext.value.errors[error];
    },
    show: (msgId, errors) => {
      errorContext.setValue(
        errors
          ? { hasErrors: true, messageId: msgId, errors: errors }
          : { hasErrors: true, messageId: msgId }
      );
    }
  };
};
