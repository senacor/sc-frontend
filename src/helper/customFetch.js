import store from '../store';
import { LOGIN_UNAUTHORIZED } from './dispatchTypes';
/**
 * supported usage:
 *   fetch('URL', configObj);
 *
 *   where configObj is something like:
 *   {
 *     method,
 *     headers,
 *     body
 *   }
 */
export default function customFetch(url, config = {}) {
  let access_token = localStorage.getItem('access_token');

  let authenticationConfig = Object.assign({}, config);
  authenticationConfig.headers = Object.assign({}, config.headers, {
    Authorization: `Bearer ${access_token}`,
    'Content-Type': 'application/json'
  });
  return fetch(url, authenticationConfig).then(response => {
    if (response.status === 401) {
      store.dispatch({
        type: LOGIN_UNAUTHORIZED,
        payload: response.status
      });
    }

    return response;
  });
}

export function fileFetch(url, config = {}) {
  let access_token = localStorage.getItem('access_token');

  let authenticationConfig = Object.assign({}, config);
  authenticationConfig.headers = Object.assign({}, config.headers, {
    Authorization: `Bearer ${access_token}`
  });
  return fetch(url, authenticationConfig).then(response => {
    if (response.status === 401) {
      store.dispatch({
        type: LOGIN_UNAUTHORIZED,
        payload: response.status
      });
    }

    return response;
  });
}
