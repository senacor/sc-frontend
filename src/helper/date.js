import moment from 'moment';
import 'moment-timezone';

export const FRONTEND_DATE_FORMAT = 'DD.MM.YYYY';
export const FRONTEND_LOCALE_DATE_TIME_FORMAT = 'YYYY-MM-DD';

export const formatDateForFrontend = input => {
  if (input) {
    let date = moment.isMoment(input) ? input : moment(input);
    if (date.isValid()) {
      return date.format(FRONTEND_DATE_FORMAT);
    }
  }
  return null;
};

// Converting LocaleDateTime for frontend
export const formatLocaleDateTime = (input, format) => {
  if (input) {
    let modifiedInput = [...input];
    modifiedInput[1] = modifiedInput[1] - 1;
    let date = moment.isMoment(modifiedInput)
      ? modifiedInput
      : moment.utc(modifiedInput);
    if (date.isValid()) {
      return date.tz('Europe/Berlin').format(format);
    }
  }
  return null;
};

// Converting date to LocaleDateTime for backend
export const convertLocalDateTime = date => {
  const convertedDate = moment
    .tz(`${date.target.value}`, 'Europe/Berlin')
    .utc()
    .format('YYYY-MM-DDTHH:mm');
  if (convertedDate === 'Invalid date') {
    return null;
  } else return convertedDate;
};

export const isValidDate = date => {
  const isValidDate = moment(
    date,
    FRONTEND_LOCALE_DATE_TIME_FORMAT,
    true
  ).isValid();
  return isValidDate;
};

export const getReadableDate = date => {
  const minutes = date[4] > 9 ? date[4].toString() : `0${date[4]}`;
  return `${date[2]}.${date[1]}.${date[0]}, ${date[3]}:${minutes}`;
};
