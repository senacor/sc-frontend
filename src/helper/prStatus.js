export const prStatusEnum = {
  RELEASED_SHEET_REVIEWER: 'FILLED_SHEET_REVIEWER',
  RELEASED_SHEET_EMPLOYEE: 'FILLED_SHEET_EMPLOYEE',
  REQUESTED_DATE: 'REQUESTED_DATE',
  FIXED_DATE: 'ALL_DATES_ACCEPTED',
  FINALIZED_REVIEWER: 'MODIFICATIONS_ACCEPTED_REVIEWER',
  FINALIZED_EMPLOYEE: 'MODIFICATIONS_ACCEPTED_EMPLOYEE',
  ARCHIVED_HR: 'PR_COMPLETED'
};
