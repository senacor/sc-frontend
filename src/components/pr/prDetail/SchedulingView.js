import React, { useContext, useEffect, useState } from 'react';
import MeetingDetailsView from './MeetingDetailsView';
import MeetingCreator from '../../scheduling/MeetingCreator';
import { prStatusEnum } from '../../../helper/prStatus';
import { injectIntl } from 'react-intl';
import {
  MeetingContext,
  UserinfoContext,
  PrContext,
  ErrorContext
} from '../../App';
import { addPrStatus } from '../../../calls/pr';

const SchedulingView = ({ pr, intl }) => {
  const { userroles, userinfo } = useContext(UserinfoContext.context).value;
  const { setValue: setPr } = useContext(PrContext.context);
  const errorContext = useContext(ErrorContext.context);

  const { value: meeting } = useContext(MeetingContext.context);
  const [canRequestMeeting, setCanRequestMeeting] = useState(false);

  useEffect(
    () => {
      if (
        meeting.status === 'DECLINED' ||
        (meeting.status === 'NOT_REQUESTED' &&
          !pr.statusSet.includes(prStatusEnum.FINALIZED_REVIEWER))
      ) {
        setCanRequestMeeting(true);
      } else {
        setCanRequestMeeting(false);
      }
    },
    [meeting]
  );

  const handleChange = () => {
    setCanRequestMeeting(!canRequestMeeting);
  };

  return (
    <div id={'outer'}>
      {canRequestMeeting ? (
        <MeetingCreator
          handleChange={() => handleChange()}
          intl={intl}
          pr={pr}
        />
      ) : (
        <MeetingDetailsView
          pr={pr}
          userinfo={userinfo}
          userroles={userroles}
          click={() =>
            addPrStatus(pr, prStatusEnum.FIXED_DATE, setPr, errorContext)
          }
          handleChange={() => handleChange()}
        />
      )}
    </div>
  );
};

export default injectIntl(SchedulingView);
