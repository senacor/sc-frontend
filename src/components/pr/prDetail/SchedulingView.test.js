import React from 'react';
import { shallow } from 'enzyme';
import { SchedulingView } from './SchedulingView';
import { prStatusEnum } from '../../../helper/prStatus';

describe('SchedulingView', () => {
  let pr = {
    id: 1,
    employee: {
      id: 2,
      login: 'test.pr.mitarbeiter1',
      firstName: 'Michaela',
      lastName: 'Mitarbeiterin'
    },
    supervisor: {
      id: 1,
      login: 'test.pr.vorgesetzter',
      firstName: 'Volker',
      lastName: 'Vorgesetzter'
    },
    statuses: [prStatusEnum.REQUESTED_DATE]
  };

  const meeting = {
    start: '2018-09-11T15:18:00.000+0000',
    end: '2018-09-11T16:18:00.000+0000',
    location: null,
    requiredAttendees: {
      'test.pr.vorgesetzter': {
        name: 'PR Vorgesetzter',
        email: 'test.pr.vorgesetzter@senacor.com',
        status: 'UNKNOWN'
      },
      'test.pr.mitarbeiter1': {
        name: 'PR Mitarbeiter1',
        email: 'test.pr.mitarbeiter1@senacor.com',
        status: 'UNKNOWN'
      }
    },
    optionalAttendees: {},
    status: 'NO_ANSWER',
    _links: {
      self: {
        href: 'http://localhost:8010/api/v1/prs/1/meetings'
      }
    }
  };

  const emptyMeeting = {
    status: 'NOT_REQUESTED'
  };

  const fetchMeetingMock = jest.fn();

  beforeEach(() => {
    fetchMeetingMock.mockClear();
  });

  it('should match snapshot', () => {
    let component = shallow(
      <SchedulingView prDetail={pr} meeting={emptyMeeting} />
    );

    expect(component).toMatchSnapshot();
  });

  it('should display subcomponent MeetingCreator if meeting is null', () => {
    let component = shallow(
      <SchedulingView prDetail={pr} meeting={emptyMeeting} />
    );

    expect(component.find('Connect(MeetingCreator)')).toHaveLength(1);
  });

  it('should display subcomponent MeetingViewDetails if meeting is not null', () => {
    let component = shallow(<SchedulingView prDetail={pr} meeting={meeting} />);

    expect(component.find('WithStyles(MeetingDetailsView)')).toHaveLength(1);
  });
});
