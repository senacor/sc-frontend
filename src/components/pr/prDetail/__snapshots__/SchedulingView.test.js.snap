// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SchedulingView should match snapshot 1`] = `
<div
  id="outer"
>
  <InjectIntl(Connect(MeetingCreator))
    handleChange={[Function]}
  />
</div>
`;
